src/java.base/share/classes/java/lang/invoke/MethodHandleNatives.java

```diff
@@ -350,10 +350,10 @@ static Object linkDynamicConstantImpl(Class<?> caller,
     }
 
     private static String staticArglistForTrace(Object staticArguments) {
-        if (staticArguments instanceof Object[])
-            return "BSA="+java.util.Arrays.asList((Object[]) staticArguments);
-        if (staticArguments instanceof int[])
-            return "BSA@"+java.util.Arrays.toString((int[]) staticArguments);
+        if (staticArguments instanceof Object[] array)
+            return "BSA="+java.util.Arrays.asList(array);
+        if (staticArguments instanceof int[] array)
+            return "BSA@"+java.util.Arrays.toString(array);
         if (staticArguments == null)
             return "BSA0=null";
         return "BSA1="+staticArguments;
```

Comment:
Nice observation and suggestions that I've totally missed, but unfortunately these code is used to set up boostrap methods, so pattern matching switches, which are implemented with bootstrap method [`SwitchBootstraps::typeSwitch`](https://github.com/openjdk/jdk/blob/4f5829e3433270d21f75ee949fbc056ae309a494/src/java.base/share/classes/java/lang/runtime/SwitchBootstraps.java#L144), cannot be used or it will lead to infinite loops like using lambdas in early startup code.

Meanwhile, even though Supplier, Consumer etc. are loaded as collections are loaded very early in startup, lambdas in their implementation methods are not called until j.l.invoke is ready, so they are safe. It's just some early startup shenanigans. Same story for early-startup records: https://github.com/openjdk/jdk/blob/4f5829e3433270d21f75ee949fbc056ae309a494/src/java.base/share/classes/jdk/internal/reflect/ReflectionFactory.java#L574-L578